dom/observable/tentative/observable-ctor.any.js

```diff
@@ -0,0 +1,685 @@
+test(() => {
+  assert_true(
+    typeof Observable === "function",
+    "Observable constructor is defined"
+  );
+}, "Observable constructor");
+
+test(() => {
+  let initializerCalled = false;
+  const results = [];
+
+  const source = new Observable((subscriber) => {
+    initializerCalled = true;
+    subscriber.next(1);
+    subscriber.next(2);
+    subscriber.next(3);
+    subscriber.complete();
+  });
+
+  assert_false(
+    initializerCalled,
+    "initializer should not be called by construction"
+  );
+
+  source.subscribe({
+    next: (x) => results.push(x),
+    error: () => assert_unreached("should not be called"),
+    complete: () => results.push("complete"),
+  });
+
+  assert_true(initializerCalled, "initializer should be called by subscribe");
+  assert_array_equals(
+    results,
+    [1, 2, 3, "complete"],
+    "should emit values synchronously"
+  );
+}, "Observable constructor calls initializer on subscribe");
+
+test(() => {
+  const error = new Error("error");
+  const results = [];
+
+  const source = new Observable((subscriber) => {
+    subscriber.next(1);
+    subscriber.next(2);
+    subscriber.error(error);
+  });
+
+  source.subscribe({
+    next: () => assert_unreached("should not be called"),
+    error: (e) => results.push(e),
+    complete: () => assert_unreached("should not be called"),
+  });
+
+  assert_array_equals(
+    results,
+    [1, 2, error],
+    "should emit error synchronously"
+  );
+}, "Observable error path called synchronously");
+
+test(() => {
+  const error = new Error("error");
+  const results = [];
+
+  const source = new Observable((subscriber) => {
+    try {
+      subscriber.next(1);
+      throw error;
+    } finally {
+      assert_true(subscriber.closed, "subscriber is closed after error");
```

Comment:
Hmm, we haven't really discussed the `.closed` attribute on the `Subscriber` interface yet, so maybe we should hold off on this assertion. Can we assert other side-effects, (i.e., would `subscriber.signal.aborted` be true? Maybe not since the outer AbortController didn't actually abort this).

Maybe it would just be good to assert the teardown is called, and talk about this attribute later.

Reply:
>  (i.e., would subscriber.signal.aborted be true? Maybe not since the outer AbortController didn't actually abort this).

There's some "dead space" where there's a difference between "closed" and "signal.aborted"

Basically, the signal is aborted _after_ your `complete` handler is fired. But at the moment `subscriber.complete` is called, the subscriber has to be shut down (`closed`).

Consider the following:


```ts
const source = new Observable(subscriber => {
  subscriber.addTeardown(() => {
    console.log('torn down!');
  });
  
  while (subscriber.signal.aborted) {
    subscriber.complete();
  }
})

const ac = new AbortController();
source.subscribe({
  complete: () => console.log('complete!'),
  signal: ac.signal
})
```

In the above case, the provided `subscriber.signal`, which is following whatever signal was passed into the `subscribe` call, would not be "aborted" yet, because it should semantically abort AFTER the `complete` handler fires. So we'll get stuck in a loop.

There's already a pieces of state we have to track internally in the `Subscriber` so that `next`, `error`, and `complete` can't be called if `error` or `complete` have already been called. This is an important part of the guarantee.

So the lifecycle inside the subscriber (if `complete` or `error` is called):

1. "close" the subscriber to new calls/actions (disable `next`, report errors passed to `error`, disable `complete`, and put `addTeardown` in a state of "execute immediately". This is a safety mechanism and a valuable guarantee.
2. emit `complete` (or `error`)
3. Tear down (by aborting the signal attached to this, which also controls everything registed with `addTeardown`.

The `closed` property of `Subscriber` exposes this "close" state from 1 above.

Reply:
I _think_ that makes sense, I just think we should put this in the explainer with a PR first, with the justification above etc. Also I think it's worth having some discussion about:

> In the above case, the provided subscriber.signal, which is following whatever signal was passed into the subscribe call, would not be "aborted" yet, because it should semantically abort AFTER the complete handler fires.

It sounds like you're saying that even with `.closed` existing, you think the AbortSignal should be aborted after `complete()` or `error()` is fired, which seems weird to me. I think only AbortController should be able to abort an AbortSignal. But maybe you're only saying this in the theoretical world where `.closed` didn't exist (like the example you have above)... either way, all of this discussion should probably happen on https://github.com/WICG/observable instead of here, if that makes sense.

Reply:
> So we'll get stuck in a loop.

This part I'm not sure I see. It seems like the re-entrant (i.e., the most nested) call to `complete()` would just trigger the outermost `complete()` handler, right? I don't think I see an actual infinite loop here.

Reply:
>  you think the AbortSignal should be aborted after complete() or error() is fired, which seems weird to me. I think only AbortController should be able to abort an AbortSignal

There's the AbortSignal that was passed into the subscription, and then there's an AbortSignal related to the status of the subscription. If the subscription is "complete" or "errors" we need to teardown resources and cancel any upcoming work.  Therefor, if you `complete()` or `error()` the `signal` that is on the `Subscriber` should be aborted. This signal would not be the same signal that was passed into the subscription.

```ts
new Observable(subscriber => {
  console.log(subscriber.signal.aborted); // false
  subscriber.complete();
  console.log(subscriber.signal.aborted); // true.
})
```